### PR TITLE
ANN: Detect E0118: Can impl only `struct`s, `enum`s and `union`s

### PR DESCRIPTION
**Pull request by mibac138**
Related #886

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 1]
```diff
@@ -864,6 +864,16 @@ sealed class RsDiagnostic(
         }
     }
 
+    class ImplForNonAdtError(
+        element: PsiElement
+    ) : RsDiagnostic(element) {
+        override fun prepare() = PreparedAnnotation(
+            ERROR,
+            E0118,
+            "Can impl only `struct`s, `enum`s and `union`s"
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: The message differs from the compiler one: `error[E0118]: no base type found for inherent implementation`

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -354,6 +355,18 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
         }
     }
 
+    // E0118: Can impl only `struct`s, `enum`s and `union`s (when not implementing a trait)
+    private fun checkImplForNonAdtError(holder: AnnotationHolder, impl: RsImplItem) {
+        if (impl.`for` != null) return
+        val typeRef = impl.typeReference ?: return
+        if (typeRef.traitType != null) return
+        val type = typeRef.type
+        if (type !is TyAdt && type != TyUnknown) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: There is an exception for primitive types: it's possible to make a single impl for primitive type with corresponding lang item, e.g. `#[lang = "u8"] impl u8 {}`. Such impls exist in `libcore`, but it's possible to make them in project in `#![no_core]` mode.

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt (hunk 1 above)
Comment (mibac138) [reply src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: I thought this message might be somewhat simpler to understand than the compiler's one and seeing how other error messages don't follow compiler error messages that closely I thought it'd be okay. I can change it to the compiler's message though no problem.

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt (hunk 2 above)
Comment (mibac138) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: Ooh, that's a good catch! I didn't really know about it. Thanks! :)

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: Okaay, sounds reasonable

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt (hunk 2 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: I just found that it's also legal to implement trait objects:
```rust
trait Bar {}
impl dyn Bar {
    fn foo(&self) {}
}
```

Comment (mibac138) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: I believe that's already covered and won't report an error due to this bit of code: `if (typeRef.traitType != null) return`. I will double check just to be sure though

Comment (vlad20012) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: `typeRef.traitType != null` is about
```rust
impl Foo for Bar {}
    //^ this trait
```
but not for the case of trait object impl

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 3]
```diff
@@ -864,6 +866,23 @@ sealed class RsDiagnostic(
...
 
+    class ImplForNonAdtError(
+        element: PsiElement,
+        private val impl: RsImplItem,
+        private val type: Ty
+    ) : RsDiagnostic(element) {
+        override fun prepare() = PreparedAnnotation(
+            ERROR,
+            E0118,
+            "Can impl only `struct`s, `enum`s and `union`s (and primitives in `no_core` crates)",
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: I don't think it's reasonable to mention primitives here, it isn't useful for must users. Mention trait objects looks much more reasonable

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 4]
```diff
@@ -864,6 +866,23 @@ sealed class RsDiagnostic(
...
+            fixes = if (type is TyPrimitive && NO_CORE.availability(element) == FeatureAvailability.CAN_BE_ADDED) {
+                listOf(NO_CORE.addFeatureFix(element))
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: No-core is not only feature, but crate attribute `#[no_core]`, and I also think that this fix isn't reasonable

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 5]
```diff
@@ -354,6 +355,21 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
         }
     }
 
+    // E0118: Can impl only `struct`s, `enum`s and `union`s (when not implementing a trait)
+    private fun checkImplForNonAdtError(holder: AnnotationHolder, impl: RsImplItem) {
+        if (impl.`for` != null) return
+        val typeRef = impl.typeReference ?: return
+        if (typeRef.traitType != null) return
+        val type = typeRef.type
+        if (type is TyPrimitive && impl.queryAttributes.langAttribute == type.name) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: :+1:

**Comment by vlad20012** [comment]
Please, rebase it to the master

**Comment by mibac138** [comment]
@vlad20012 done

**Comment by mchernyavsky** [comment]
Thanks!
bors r+

**Pull request merged**